apps/src/lib/tools/jsdebugger/DebugConsole.jsx

```diff
@@ -0,0 +1,187 @@
+import React, {PropTypes} from 'react';
+import {connect} from 'react-redux';
+
+import {KeyCodes} from '../../../constants';
+import {
+  add as addWatchExpression,
+  remove as removeWatchExpression
+} from '../../../redux/watchedExpressions';
+import CommandHistory from './CommandHistory';
+import JSInterpreter from '../../../JSInterpreter';
+
+import {actions, selectors} from './redux';
+
+const WATCH_COMMAND_PREFIX = "$watch ";
+const UNWATCH_COMMAND_PREFIX = "$unwatch ";
+
+/**
+ * Set the cursor position to the end of the text content in a div element.
+ * @see http://stackoverflow.com/a/6249440/5000129
+ * @param {!HTMLDivElement} element
+ */
+function moveCaretToEndOfDiv(element) {
+  var range = document.createRange();
+  if (element.childNodes.length === 0) {
+    return;
+  }
+
+  range.setStart(element.lastChild, element.lastChild.nodeValue.length);
+  range.collapse(true);
+
+  // Change window selection to new range to set cursor position
+  var selection = window.getSelection();
+  selection.removeAllRanges();
+  selection.addRange(range);
+}
+
+
+/**
+ * The console for our debugger UI
+ */
+const DebugConsole = connect(
+  state => ({
+    commandHistory: selectors.getCommandHistory(state),
+    jsInterpreter: selectors.getJSInterpreter(state),
+    logOutput: selectors.getLogOutput(state),
+  }),
+  {
+    addWatchExpression,
+    removeWatchExpression,
+    appendLog: actions.appendLog,
+  }
+)(React.createClass({
+  propTypes: {
+    // from redux
+    commandHistory: PropTypes.instanceOf(CommandHistory),
+    jsInterpreter: PropTypes.instanceOf(JSInterpreter),
+    logOutput: PropTypes.string.isRequired,
+    addWatchExpression: PropTypes.func.isRequired,
+    removeWatchExpression: PropTypes.func.isRequired,
+    appendLog: PropTypes.func.isRequired,
+
+    // passed from above
+    debugButtons: PropTypes.bool,
+    debugWatch: PropTypes.bool,
+    style: PropTypes.object,
+  },
+
+  onInputKeyDown(e) {
+    var input = e.target.textContent;
+    if (e.keyCode === KeyCodes.ENTER) {
+      e.preventDefault();
+      this.props.commandHistory.push(input);
+      e.target.textContent = '';
+      this.appendLog('> ' + input);
+      if (0 === input.indexOf(WATCH_COMMAND_PREFIX)) {
+        this.props.addWatchExpression(
+          input.substring(WATCH_COMMAND_PREFIX.length)
+        );
+      } else if (0 === input.indexOf(UNWATCH_COMMAND_PREFIX)) {
+        this.props.removeWatchExpression(
+          input.substring(UNWATCH_COMMAND_PREFIX.length)
+        );
+      } else if (this.props.jsInterpreter) {
```

Comment:
I like the state breakdown  (uninitialized, attached, paused, etc) in the description of https://github.com/code-dot-org/code-dot-org/pull/13247. I would recommend adding a helper function getInterpreterState or similar which returns one of the 4 values, rather than checking for the existence of the interpreter here, to make the current state more visible. However, I'm open to arguments against.